### PR TITLE
feat(gh-299): Type hints in distributions modules

### PR DESCRIPTION
**Pull request by Qazalbash**
Hi,

I have added type hints using [`jaxtyping`](https://github.com/patrick-kidger/jaxtyping) in `numpyro.distributions.*.py` modules. I have accordingly updated the `setup.py` too.

All types of protocols have been transferred to `numpyro._typing`. I have modified the `DistributionLike` type along with two new types, `TransformLike` and `ConstraintLike`.

---

This PR is related to #299.

### CODE REVIEW

**Comment by fehiepsi** [comment]
Just curious what jaxtyping offers?

**Comment by Qazalbash** [comment]
Sure, we can avoid it!

**Comment by Qazalbash** [comment]
All test cases are passing except for those failing on the master branch. I have figured out the problem and [reported it](https://github.com/Joshuaalbert/jaxns/issues/221).

File: numpyro/_typing.py [hunk 1]
```diff
@@ -1,14 +1,99 @@
...
+
 P = ParamSpec("P")
 ModelT: TypeAlias = Callable[P, Any]
 
 Message: TypeAlias = dict[str, Any]
 TraceT: TypeAlias = OrderedDict[str, Message]
+
+
+@runtime_checkable
+class ConstraintLike(Protocol):
```
Comment (fehiepsi) [inline numpyro/_typing.py]: nit: maybe use ConstraintT to be consistent with some other types?

Comment (Qazalbash) [reply numpyro/_typing.py]: I think it's a good suggestion, I will update the PR shortly!

File: numpyro/_typing.py [hunk 2]
```diff
@@ -1,14 +1,99 @@
...
+    pytree_data_fields: tuple = ...
+    pytree_aux_fields: tuple = ...
+
+    def __call__(self, *args: Any, **kwargs: Any) -> Any: ...
+
+    def rsample(
+        self, key: jax.dtypes.prng_key, sample_shape: tuple[int, ...] = ()
+    ) -> ArrayLike: ...
+    def sample(
+        self, key: jax.dtypes.prng_key, sample_shape: tuple[int, ...] = ()
```
Comment (fehiepsi) [inline numpyro/_typing.py:54]: maybe Optional[jax.dtypes.prng_key]?

File: numpyro/distributions/distribution.py [hunk 3]
```diff
@@ -303,13 +312,15 @@ def event_dim(self) -> int:
     def has_rsample(self) -> bool:
         return set(self.reparametrized_params) == set(self.arg_constraints)
 
-    def rsample(self, key, sample_shape=()) -> ArrayLike:
+    def rsample(
+        self, key: jax.dtypes.prng_key, sample_shape: tuple[int, ...] = ()
```
Comment (fehiepsi) [inline numpyro/distributions/distribution.py]: Optional[jax.dtypes.prng_key]

File: numpyro/distributions/distribution.py [hunk 4]
```diff
@@ -324,7 +335,9 @@ def shape(self, sample_shape=()) -> tuple[int, ...]:
         """
         return sample_shape + self.batch_shape + self.event_shape
 
-    def sample(self, key: ArrayLike, sample_shape: tuple[int, ...] = ()) -> ArrayLike:
+    def sample(
+        self, key: jax.dtypes.prng_key, sample_shape: tuple[int, ...] = ()
```
Comment (fehiepsi) [inline numpyro/distributions/distribution.py]: Optional[jax.dtypes.prng_key]

File: numpyro/distributions/distribution.py [hunk 5]
```diff
@@ -666,14 +650,21 @@ def _broadcast_shape(existing_shape, new_shape):
...
     @property
-    def has_rsample(self):
+    def has_rsample(self) -> bool:
         return self.base_dist.has_rsample
 
-    def _sample(self, sample_fn, key, sample_shape=()):
+    def _sample(
+        self,
+        sample_fn: Callable[
+            [jax.dtypes.prng_key, tuple[int, ...]], tuple[ArrayLike, list[ArrayLike]]
```
Comment (fehiepsi) [inline numpyro/distributions/distribution.py]: Optional[jax.dtypes.prng_key]

File: numpyro/distributions/distribution.py [hunk 6]
```diff
@@ -666,14 +650,21 @@ def _broadcast_shape(existing_shape, new_shape):
...
+        ],
+        key: jax.dtypes.prng_key,
```
Comment (fehiepsi) [inline numpyro/distributions/distribution.py]: Optional[jax.dtypes.prng_key]

File: numpyro/distributions/distribution.py [hunk 7]
```diff
@@ -709,24 +700,28 @@ def reshape_sample(x):
         samples = reshape_sample(samples)
         return samples, intermediates
 
-    def rsample(self, key, sample_shape=()):
+    def rsample(self, key: jax.dtypes.prng_key, sample_shape: tuple[int, ...] = ()):
```
Comment (fehiepsi) [inline numpyro/distributions/distribution.py]: Optional[jax.dtypes.prng_key]

File: numpyro/distributions/distribution.py [hunk 8]
```diff
@@ -709,24 +700,28 @@ def reshape_sample(x):
...
         )[0]
 
     @property
-    def support(self):
+    def support(self) -> constraints.Constraint:
         return self.base_dist.support
 
-    def sample_with_intermediates(self, key, sample_shape=()):
+    def sample_with_intermediates(
+        self, key: jax.dtypes.prng_key, sample_shape: tuple[int, ...] = ()
```
Comment (fehiepsi) [inline numpyro/distributions/distribution.py]: Optional[jax.dtypes.prng_key]

File: numpyro/distributions/distribution.py [hunk 9]
```diff
@@ -709,24 +700,28 @@ def reshape_sample(x):
...
+    ) -> tuple[ArrayLike, list[ArrayLike]]:
         return self._sample(self.base_dist.sample_with_intermediates, key, sample_shape)
 
-    def sample(self, key, sample_shape=()):
+    def sample(
+        self, key: jax.dtypes.prng_key, sample_shape: tuple[int, ...] = ()
```
Comment (fehiepsi) [inline numpyro/distributions/distribution.py]: Optional[jax.dtypes.prng_key]

File: numpyro/_typing.py [hunk 10]
```diff
@@ -1,14 +1,99 @@
...
+    has_enumerate_support: bool = ...
+    reparametrized_params: list[str] = ...
+    _validate_args: bool = ...
+    pytree_data_fields: tuple = ...
+    pytree_aux_fields: tuple = ...
+
+    def __call__(self, *args: Any, **kwargs: Any) -> Any: ...
+
+    def rsample(
+        self, key: jax.dtypes.prng_key, sample_shape: tuple[int, ...] = ()
```
Comment (fehiepsi) [inline numpyro/_typing.py:51]: Optional[jax.dtypes.prng_key]

File: numpyro/distributions/distribution.py [hunk 11]
```diff
@@ -338,7 +351,9 @@ def sample(self, key: ArrayLike, sample_shape: tuple[int, ...] = ()) -> ArrayLik
         """
         raise NotImplementedError
 
-    def sample_with_intermediates(self, key, sample_shape=()):
+    def sample_with_intermediates(
+        self, key: jax.dtypes.prng_key, sample_shape: tuple[int, ...] = ()
```
Comment (fehiepsi) [inline numpyro/distributions/distribution.py]: Optional[jax.dtypes.prng_key]

File: numpyro/distributions/distribution.py [hunk 12]
```diff
@@ -880,38 +888,40 @@ def __init__(self, base_dist, reinterpreted_batch_ndims, *, validate_args=None):
...
         return self.base_dist.variance
 
     @property
-    def has_rsample(self):
+    def has_rsample(self) -> bool:
         return self.base_dist.has_rsample
 
-    def rsample(self, key, sample_shape=()):
+    def rsample(self, key: jax.dtypes.prng_key, sample_shape: tuple[int, ...] = ()):
         return self.base_dist.rsample(key, sample_shape=sample_shape)
```
Comment (fehiepsi) [inline numpyro/distributions/distribution.py:922]: Optional[jax.dtypes.prng_key]

File: numpyro/distributions/distribution.py [hunk 13]
```diff
@@ -880,38 +888,40 @@ def __init__(self, base_dist, reinterpreted_batch_ndims, *, validate_args=None):
...
 
-    def sample(self, key, sample_shape=()):
+    def sample(
+        self, key: jax.dtypes.prng_key, sample_shape: tuple[int, ...] = ()
```
Comment (fehiepsi) [inline numpyro/distributions/distribution.py]: Optional[jax.dtypes.prng_key]

File: numpyro/distributions/distribution.py [hunk 14]
```diff
@@ -959,24 +969,26 @@ def __init__(self, base_dist, mask):
...
+    def has_enumerate_support(self) -> bool:
         return self.base_dist.has_enumerate_support
 
     @property
-    def has_rsample(self):
+    def has_rsample(self) -> bool:
         return self.base_dist.has_rsample
 
-    def rsample(self, key, sample_shape=()):
+    def rsample(self, key: jax.dtypes.prng_key, sample_shape: tuple[int, ...] = ()):
```
Comment (fehiepsi) [inline numpyro/distributions/distribution.py]: Optional[jax.dtypes.prng_key]

File: numpyro/distributions/distribution.py [hunk 15]
```diff
@@ -959,24 +969,26 @@ def __init__(self, base_dist, mask):
...
         return self.base_dist.rsample(key, sample_shape=sample_shape)
 
     @property
     def support(self):
         return self.base_dist.support
 
-    def sample(self, key, sample_shape=()):
+    def sample(
+        self, key: jax.dtypes.prng_key, sample_shape: tuple[int, ...] = ()
```
Comment (fehiepsi) [inline numpyro/distributions/distribution.py]: Optional[jax.dtypes.prng_key]

File: numpyro/distributions/distribution.py [hunk 16]
```diff
@@ -1103,10 +1121,10 @@ def __init__(self, base_distribution, transforms, *, validate_args=None):
         )
 
     @property
-    def has_rsample(self):
+    def has_rsample(self) -> bool:
         return self.base_dist.has_rsample
 
-    def rsample(self, key, sample_shape=()):
+    def rsample(self, key: jax.dtypes.prng_key, sample_shape: tuple[int, ...] = ()):
```
Comment (fehiepsi) [inline numpyro/distributions/distribution.py]: Optional[jax.dtypes.prng_key]

File: numpyro/distributions/distribution.py [hunk 17]
```diff
@@ -1124,13 +1142,17 @@ def support(self):
                 codomain, self.event_dim - codomain_event_dim
             )
 
-    def sample(self, key, sample_shape=()):
+    def sample(
+        self, key: jax.dtypes.prng_key, sample_shape: tuple[int, ...] = ()
```
Comment (fehiepsi) [inline numpyro/distributions/distribution.py]: Optional[jax.dtypes.prng_key]

File: numpyro/distributions/distribution.py [hunk 18]
```diff
@@ -1124,13 +1142,17 @@ def support(self):
...
+    ) -> ArrayLike:
         x = self.base_dist(rng_key=key, sample_shape=sample_shape)
         for transform in self.transforms:
             x = transform(x)
         return x
 
-    def sample_with_intermediates(self, key, sample_shape=()):
+    def sample_with_intermediates(
+        self, key: jax.dtypes.prng_key, sample_shape: tuple[int, ...] = ()
```
Comment (fehiepsi) [inline numpyro/distributions/distribution.py]: Optional[jax.dtypes.prng_key]

File: numpyro/distributions/distribution.py [hunk 19]
```diff
@@ -1242,27 +1273,29 @@ def __init__(self, v=0.0, log_density=0.0, event_dim=0, *, validate_args=None):
     def support(self):
         return constraints.independent(constraints.real, self.event_dim)
 
-    def sample(self, key, sample_shape=()):
+    def sample(
+        self, key: jax.dtypes.prng_key, sample_shape: tuple[int, ...] = ()
```
Comment (fehiepsi) [inline numpyro/distributions/distribution.py]: Optional[jax.dtypes.prng_key]

File: numpyro/distributions/distribution.py [hunk 20]
```diff
@@ -571,43 +588,6 @@ def __repr__(self) -> str:
         )
 
 
-@runtime_checkable
-class DistributionLike(Protocol):
```
Comment (fehiepsi) [inline numpyro/distributions/distribution.py:575]: Maybe set DistributionLike = DistributionT here to avoid breaking users' code?

**Review by fehiepsi — APPROVED** [review]
LGTM pending the usage of Optional[jax.dtypes.prng_key] at some places.

**Review by fehiepsi — APPROVED** [review]

File: numpyro/_typing.py (hunk 2 above)
Comment (Qazalbash) [reply numpyro/_typing.py:54]: I set it to optional only where we can pass `None`, like in `Unit` distribution. Setting it to `Optional[jax.dtypes.prng_key]` will give an illusion that `None` can also be passed. What do you say?

Comment (fehiepsi) [reply numpyro/_typing.py:54]: I see, so subclasses can define new argument types without violating Protocol? Please disregard my comments for this and Distribution class. Other classes need to support None type I think.

Comment (Qazalbash) [reply numpyro/_typing.py:54]: > I see, so subclasses can define new argument types without violating Protocol?

Mypy is also passing them, which means it should be correct.

> Please disregard my comments for this and Distribution class. Other classes need to support None type I think.

I think we should make it consistent with `Optional[jax.dtypes.prng_key]`, it will avoid confusion.

Comment (fehiepsi) [reply numpyro/_typing.py:54]: It is up to you to decide how to deal with base classes. Just want to make sure that those Delta, Expanded, Transformed, Masked should support None value.
